server/api/cvrs.py

```diff
@@ -782,20 +783,93 @@ def join_ballots_to_interpretations(
-    # 2. First, parse out the contest and choice names. We have to do this in
-    #    a separate pass since our storage scheme for interpretations requires
-    #    knowing all of the contest and choice names up front.
-    # 3. Next, parse out the interpretations.
+    # Hart CVRs consist of a ZIP file containing an individual XML file for each ballot's CVR.
+    # Separate from this ZIP file, an optional scanned ballot information CSV can be provided. If
+    # the latter is provided, the `UniqueIdentifier`s in it will be used as imprinted IDs.
+    # Otherwise, `CvrGuid`s will be used.
+    #
+    # Our parsing steps:
+    # 1. Unzip the wrapper ZIP file.
+    # 2. Expect either [ another ZIP file ] or [ another ZIP file and a CSV ].
+    # 3. If a CSV is found, parse it as a scanned ballot information CSV.
+    # 4. Unzip the CVRs ZIP file.
+    # 5. Parse out the contest and choice names. We have to do this in a separate pass since our
+    #    storage scheme for interpretations requires knowing all of the contest and choice names up
+    #    front.
+    # 3. Then, parse out the interpretations.
+
+    wrapper_zip_file = retrieve_file(jurisdiction.cvr_file.storage_path)
+    files = unzip_files(wrapper_zip_file)
+    wrapper_zip_file.close()
+
+    cvrs_zip_file: Union[BinaryIO, None] = None
+    scanned_ballot_information_file: Union[BinaryIO, None] = None
+    for name, file in files.items():
+        if name.lower().endswith(".zip"):
+            cvrs_zip_file = file
+        if name.lower().endswith(".csv"):
+            scanned_ballot_information_file = file
+
+    if cvrs_zip_file is None:
+        raise UserError("Expected a file with a .zip extension.")
+
+    def construct_cvr_guid_to_unique_identifier_mapping(
+        scanned_ballot_information_file: BinaryIO,
+    ) -> Dict[str, str]:
+        validate_not_empty(scanned_ballot_information_file)
+        text_file = decode_csv(scanned_ballot_information_file)
+
+        # Determine whether the scanned ballot information CSV has a format version row and skip it
+        # if so
+        has_format_version_row = False
+        first_line = text_file.readline()
+        if first_line.startswith("#FormatVersion") or "," not in first_line:
+            has_format_version_row = True
+            validate_comma_delimited(text_file)
+            # validate_comma_delimited resets the cursor to the start of the file so skip the
+            # format version row again
+            text_file.readline()
+        else:
+            text_file.seek(0)
+            validate_comma_delimited(text_file)
+        reader = csv.reader(text_file, delimiter=",")
+
+        headers = next(reader)
+        if "CvrId" not in headers:
+            raise UserError("Scanned ballot information CSV missing CvrId column.")
+        if "UniqueIdentifier" not in headers:
+            raise UserError(
+                "Scanned ballot information CSV missing UniqueIdentifier column."
+            )
```

Comment:
Check out how the `get_header_indices` and `column_value` functions are used elsewhere in this file. They handle a lot of the edge cases and error handling needed for parsing out CSV values. That should make some of this code a bit simpler and more consistent with the rest of the CVR parsing code.

Reply:
Nice! Thanks for pointing these out - https://github.com/votingworks/arlo/pull/1558/commits/d599ae4b63474ef008250428746a5057b87ad84c